compiler/rustc_codegen_llvm/src/debuginfo/metadata/enums/cpp_like.rs

```diff
@@ -532,21 +532,39 @@ fn build_variant_struct_wrapper_type_di_node<'ll, 'tcx>(
                 variant_struct_type_di_node,
             ));
 
-            let build_assoc_const =
-                |name: &str, type_di_node: &'ll DIType, value: u64, align: Align| unsafe {
-                    llvm::LLVMRustDIBuilderCreateStaticMemberType(
-                        DIB(cx),
-                        wrapper_struct_type_di_node,
-                        name.as_c_char_ptr(),
-                        name.len(),
-                        unknown_file_metadata(cx),
-                        UNKNOWN_LINE_NUMBER,
-                        type_di_node,
-                        DIFlags::FlagZero,
-                        Some(cx.const_u64(value)),
-                        align.bits() as u32,
-                    )
+            let build_assoc_const = |name: &str,
+                                     type_di_node_: &'ll DIType,
+                                     value: u64,
+                                     align: Align| unsafe {
+                // FIXME: Currently we force all DISCR_* values to be u64's as LLDB seems to have
+                // problems inspecting other value types. Since DISCR_* is typically only going to be
+                // directly inspected via the debugger visualizer - which compares it to the `tag` value
+                // (whose type is not modified at all) it shouldn't cause any real problems.
+                let (t_di, align) = if name == ASSOC_CONST_DISCR_NAME {
+                    (type_di_node_, align.bits() as u32)
+                } else {
+                    let ty_u64 = Ty::new_uint(cx.tcx, ty::UintTy::U64);
+                    (type_di_node(cx, ty_u64), Align::EIGHT.bits() as u32)
                 };
+                const DW_TAG_CONST_TYPE: c_uint = 0x0026;
+
+                // must wrap type in a `const` modifier for LLDB to be able to inspect the value of the member
+                let field_type =
+                    llvm::LLVMRustDIBuilderCreateQualifiedType(DIB(cx), DW_TAG_CONST_TYPE, t_di);
```

Comment:
Discussion: this is really 🌟 magic 🌟, would a better approach be to properly teach lldb how to handle this...? Do we want to maintain this kind of hack? (I'm not a debuginfo expert by any means, just observing that this seems quite hacky)

Reply:
People are unfortunately often using wildly dated debuggers, even if they use a more recent compiler, so we should probably do this anyways.

And as far as debuginfo hacks, this is pretty tame.

Reply:
Adding the const modifier via `QualifiedType` is more or less how things are supposed to be done as far as I can tell. `cl.exe` outputs PDB info with the `QualifiedType` wrapper for C/C++ code, so Rust probably should as well. The values *are* constants, so there's no reason they shouldn't be marked as such. The fact that microsoft's debugger and DWARF debuggers can inspect non-const static fields seems like the anomaly, since non-const static fields in C/C++ can't have inline initializers in the first place. 

As for forcing the type of the `DISCR_*`, I don't think it's any more of a hack than the `DISCR_*` value itself - it's a "made up" value meant to bypass the fact that `TypeSystemClang` doesn't really support sum types or niche optimization. The only place `DISCR_*` is used is in [lldb_providers.py, gdb_providers.py, and intrinsic.natvis](https://github.com/rust-lang/rust/tree/master/src/etc) to make it possible to determine which enum variant the `tag` value corresponds to. For practical purposes, it doesn't really matter what type `DISCR_*` is, so long as it can be compared to `tag` in debugger visualizer scripts.

After a bit of extra fiddling, it seems to exclusively have problems with `unsigned char`. I have absolutely no idea, but it fails to do read them for `cl.exe`-compiled C/C++ as well. I'm not familiar enough with LLVM internals to know where to even start looking for a solution to that.

Reply:
Fair enough, thanks for the elaboration 👍